string side
float32 distance